atomic_reactor/utils/resource_manager.py

```diff
@@ -67,186 +166,378 @@ def username(self):
     def ssh_keyfile(self):
         return self._ssh_keyfile
 
-    def _validate_resource_lock(self, payload: dict):
-        """ Validate resource lock json file against schema """
-        with open(RESOURCE_LOCK_SCHEMA, 'r') as f:
-            schema_data = f.read()
-        schema = json.loads(schema_data)
+    @property
+    def slots(self):
+        return self._slots
+
+    @cached_property
+    def slots_dir(self):
+        home, _, _ = self._run("pwd")
+        return os.path.join(home, SLOTS_RELATIVE_PATH)
+
+    def _is_valid_slot_id(self, slot_id: int) -> bool:
+        """ Check if a slot id is valid """
+        valid_slots = list(range(self.slots))
+        if slot_id not in valid_slots:
+            logger.error("%s: invalid slot id %s, should be in: %s",
+                         self.hostname, slot_id, valid_slots)
+            return False
+        return True
+
+    def _get_slot_path(self, slot_id: int) -> str:
+        """ Get the absolute path of slot file """
+        return os.path.join(self.slots_dir, f"slot_{slot_id}")
+
+    def _get_slot_lock_path(self, slot_id: int) -> str:
+        """ Get the absolute path of slot's lock file """
+        return os.path.join(self.slots_dir, f"slot_{slot_id}.lock")
+
+    @backoff.on_exception(
+        backoff.expo,
+        SlotLockError,
+        factor=BACKOFF_FACTOR,
+        max_tries=MAX_RETRIES,
+        jitter=None,  # use deterministic backoff, do not apply random jitter
+        logger=logger,
+    )
+    def _get_blocking_session_with_locked_slot(
+        self, session: SSHRetrySession, slot_id: int
+    ) -> Tuple[SSHRetrySession, ChannelFile, ChannelFile, ChannelFile]:
+        """
+        Lock the slot in SSH session and keep it blocked
+
+        :param session: SSHRetrySession, an SSH session
+        :param slot_id: int, slot ID
+        :return: A tuple of session and the stdin, stdout, stderr of the running command
+        """
+        # Run `cat` in the session to keep the slot lock file being locked
+        lock_path = quote(self._get_slot_lock_path(slot_id))
+        cmd = f"flock --conflict-exit-code 42 --nonblocking {lock_path} cat"
+
+        _errmsg = f"{self.hostname}: failed to acquire lock on slot {slot_id}"
+        try:
+            logger.info("%s: acquiring lock on slot %s", self.hostname, slot_id)
+            stdin, stdout, stderr = session.exec_command(cmd)
+        except Exception as ex:
+            raise SlotLockError(_errmsg) from ex
+
+        # A short time sleep to wait for the socket to be ready
+        time.sleep(0.1)
```

Comment:
yes, it could happen when the socket is not ready for write, in my testing `time.sleep(0.01)` is not good enough, so I set it to `0.1` second and seems working without issue. And yes, you're right it also happens when slot is locked by others.

Reply:
Ack, seems reasonable